networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py

```diff
@@ -0,0 +1,255 @@
+
+
+class AciASR1kRoutingDriver(asr1k.ASR1kRoutingDriver):
+
+    def __init__(self, **device_params):
+        super(AciASR1kRoutingDriver, self).__init__(**device_params)
+        self._fullsync = False
+        self._deployment_id = "zxy"
+        self.hosting_device = {'id': device_params.get('id'),
+                               'device_id': device_params.get('device_id')}
+
+    # ============== Public functions ==============
+
+    def internal_network_added(self, ri, port):
+        if not self._is_port_v6(port):
+            if self._is_global_router(ri):
+                # The global router is modeled as the default vrf
+                # in the ASR.  When an external gateway is configured,
+                # a normal "internal" interface is created in the default
+                # vrf that is in the same subnet as the ext-net.
+                LOG.debug("global router handling")
+                self.external_gateway_added(ri, port)
+            else:
+                LOG.debug("Adding IPv4 internal network port: %(port)s "
+                          "for router %(r_id)s", {'port': port, 'r_id': ri.id})
+                self._create_sub_interface(
+                    ri, port, is_external=False)
+                self._add_tenant_net_route(ri, port)
+
+    def _create_sub_interface(self, ri, port, is_external=False, gw_ip=""):
+        vlan = self._get_interface_vlan_from_hosting_port(port)
+        if (self._fullsync and
+                int(vlan) in self._existing_cfg_dict['interfaces']):
+            LOG.info(_LI("Sub-interface already exists, skipping"))
+            return
+        vrf_name = self._get_vrf_name(ri)
+        hsrp_ip = self._get_interface_ip_from_hosting_port(port,
+            is_external=is_external)
+        net_mask = self._get_interface_subnet_from_hosting_port(port,
+            is_external=is_external)
+        sub_interface = self._get_interface_name_from_hosting_port(port)
+        self._do_create_sub_interface(sub_interface, vlan, vrf_name, hsrp_ip,
+                                      net_mask, is_external)
+        # Always do HSRP
+        if ri.router.get(ha.ENABLED, False):
+            if port.get(ha.HA_INFO) is not None:
+                self._add_ha_hsrp(ri, port, is_external=is_external)
+            else:
+                # We are missing HA data, candidate for retrying
+                params = {'r_id': ri.router_id, 'p_id': port['id'],
+                          'port': port}
+                raise cfg_exc.HAParamsMissingException(**params)
+
+    def _add_ha_hsrp(self, ri, port, is_external=False):
+        priority = None
+        if ri.router.get(ROUTER_ROLE_ATTR) in (ROUTER_ROLE_HA_REDUNDANCY,
+                                               ROUTER_ROLE_GLOBAL):
+            for router in ri.router[ha.DETAILS][ha.REDUNDANCY_ROUTERS]:
+                if ri.router['id'] == router['id']:
+                    priority = router[ha.PRIORITY]
```

Comment:
Good question -- this was leveraged from similar code in the stock ASR driver. I suspect that you're correct, but am going to leave this in for now... if a fix is needed, we'll create a separate patch that addresses any such issue across router type drivers.